src/Silverback.Integration/Diagnostics/BrokerHeadersEnricherFactory.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Silverback.Messaging;
+using Silverback.Messaging.Messages;
+
+namespace Silverback.Diagnostics
+{
+    internal sealed class BrokerHeadersEnricherFactory : IBrokerHeadersEnricherFactory
+    {
+        private static readonly NullEnricher NullEnricherInstance = new();
+
+        private readonly IServiceProvider _serviceProvider;
+
+        private readonly ConcurrentDictionary<Type, Type> _enricherTypeCache = new();
+
+        public BrokerHeadersEnricherFactory(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public IBrokerHeadersEnricher GetHeadersEnricher(IEndpoint endpoint)
```

Comment:
I wonder if we could/should create a base class, since this logic is now used in 3 places, right? Something like an `abstract BrokerEnricherFactory<TEnricher>`. WDYT?